Implement factory creation

weierophinney opened the pull request:
This patch ~~is a work in progress, addressing~~ addresses #42.

- [x] Functionality for creating a factory by reflecting a class
- [x] Functionality for writing a factory class file for a given class name
- [x] symfony/console command `factory:create <class>`
- [x] Update `handler:create` to also generate a factory for the created handler; command should include a `--no-factory` switch to disable this
- [x] Update `middleware:create` to also generate a factory for the created middleware; command should include a `--no-factory` switch to disable this
- [x] Register class/factory pairs in `config/autoload/dependencies.global.php`

~~One question I have: should the command be `factory:create` or `factory:create-for`?~~

The final commands are:

- `expressive factory:create <for class name> [--no-register]`
- `expressive handler:create <class name> [--no-factory] [--no-register]`
- `expressive middleware:create <class name> [--no-factory] [--no-register]`

weierophinney committed:
Adds functionality for generating a factory class

Creates and returns the string contents for the factory class based on
reflecting the given class. Factories:

- use strict mode
- are generic classes defining `__invoke`, which accepts a PSR-11 container
- import the PSR-11 container and any classes the class constructor depends on, even if they are in the same namespace
- pull class dependencies from the container

weierophinney committed:
Adds class for creating factory classes

Composes a `FactoryClassGenerator` in order to produce a factory for the
given class name, and then writes that class in a sibling file.

If the class is not autoloadable, `Create` raises an exception.

If unable to write the file to the filesystem, `Create` also raises an
exception.

The factory method, `createForClass`, returns the filename where the
factory class was written.

weierophinney committed:
Adds symfony/console command class `CreateFactoryCommand`

weierophinney committed:
Registers factory:create command with CLI tool

weierophinney committed:
Also raise an exception if the factory file to generate already exists

geerteltink commented:
> One question I have: should the command be factory:create or factory:create-for?

`factory:create` sounds better and is shorter

svycka commented:
> It would be useful if the create middleware functionality could add a skeleton factory and register it in the module's ConfigProvider.
> Where should the factory be saved? I've seen people saving it in the same dir and others in a special Factory dir.
> > Same directory. As this is what the skeleton does for HomePageFactory. If people put it elsewhere, they can not use this tool or move it.

maybe add config option for this? or ask for the first time or even more fun it could check if exists Factory directory then create there instead

emir commented:
`factory:create` can be predictable and keeps consistency to other commands like `handler:create`, `middleware:create`

weierophinney commented:
> maybe add config option for this?

This hugely complicates the command, and makes errors during generation far more likely. The only way it would work reliably is if, in addition to the class name for which to generate the factory, we required **both** the class name for the new factory, **and** the path where the factory should be created. At that point, you're typing as much, if not more, than you would if you were to generate the factory and then perform a `git mv` operation.

weierophinney committed:
Adds class for injecting factory configuration

`ConfigInjector` creates and/or updates a special autoloadable config
file, zend-expressive-tooling-factories.global.php. Its
`injectFactoryForClass()` method expects the factory class name and the
class name to which it maps. It slurps up the existing configuration, if
any, and then adds the new mapping, sorts the entries, and writes the
file back to its location.

weierophinney committed:
Updates CreateFactoryCommand to register factory

Modifies the CreateFactoryCommand to register the new factory for the
class using the ConfigInjector by default. Users can opt to disable this
feature via the `--no-register` option.

weierophinney committed:
Updates middleware:create to allow generating a factory by default

When `middleware:create` is called, it now generates a factory for the
generated middleware, as a sibling class, using `factory:create`.

Users can disable this by passing the `--no-factory` class.

Users can also generate the factory, but not register it, by passing the
`--no-register` option.

weierophinney committed:
Updates `handler:create` to generate a factory by default

Updates the `handler:create` command to trigger the `factory:create`
command with the name of the generated handler.

Users may disable generation of the factory using the option
`--no-factory`.

Users may generate the factory, but omit registration of the
handler/factory pair by passing the option `--no-register`.

weierophinney committed:
Ensure factory generation works in actual usage

When testing the `middleware:create` and `handler:create` commands, I
received exceptions during factory generation saying that the middleware
and/or handler class could not be found. Adding a `require $path`
statement addresses the issue, but introduces new problems during
testing, which I addressed by adding a private flag that, when disabled,
skips the `require` statement.

geerteltink commented on src/Factory/ConfigInjector.php:
```diff
@@ -0,0 +1,115 @@
+ */
+class ConfigInjector
+{
+    public const CONFIG_FILE = 'config/autoload/zend-expressive-tooling-factories.global.php';
```
Is there a reason why a new file is used and not `config/autoload/dependencies.global.php`?

weierophinney replied:
Yes: to prevent issues when re-generating the file:

- serialization of closures
- resolution of constants
- expressions
- usage of import statements

It's far more complex to write to a file _meant_ to be managed manually than to use a file that we can mark as _under control of the tooling_.

geerteltink replied:
Makes sense. Thanx for the explanation.

weierophinney committed:
Adds CHANGELOG entry for #52

michalbundyra commented on src/CreateHandler/CreateHandlerCommand.php:
```diff
@@ -63,6 +89,22 @@ protected function execute(InputInterface $input, OutputInterface $output)
         ));
 
+        if (! $input->getOption('no-factory')) {
+            $this->requireHandlerBeforeGeneratingFactory && require $path;
```
This is unclear, not easy readable. I know how it works, but it looks like 'js optimized code'.

What's more I don't like to include the class that way. Can't we use autoloader? If not, maybe we should use BetterReflection to read the class by reflection without autoloading the class?

If there is no other way than require please change it for more readable `if` statement

weierophinney replied:
It's a testing problem, plain and simple. `require` works fine in actual use, but in testing, we get an error due to the fact that no resolvable class file is created.

I can update this to be an `if` statement; it just feels like overkill for what will be the general use case.

weierophinney replied:
And no, the autoloader does not work in this situation. I'm not entirely sure why. But I tried a number of approaches, and the only way I could get it to be able to reflect the generated class was to require it first. I am reluctant to add `BetterReflection` when we can accomplish what we need without the extra dependency.

michalbundyra commented on src/CreateMiddleware/CreateMiddlewareCommand.php:
```diff
@@ -63,6 +88,22 @@ protected function execute(InputInterface $input, OutputInterface $output)
         ));
 
+        if (! $input->getOption('no-factory')) {
+            $this->requireMiddlewareBeforeGeneratingFactory && require $path;
```
The same as noted above.

michalbundyra commented on src/Factory/ConfigInjector.php:
```diff
@@ -0,0 +1,115 @@
+            throw ConfigFileNotWritableException::forFile($this->configFile);
+        }
+
+        $config = file_exists($this->configFile) ? include($this->configFile) : [];
```
I think usually we are not using parenthesis with `include`. We have this check in upcoming zend-coding-standards

weierophinney replied:
I had done this because originally I was calculating the filename inline; I'll remove the parens in an upcoming commit.

michalbundyra commented on src/Factory/ConfigInjector.php:
```diff
@@ -0,0 +1,115 @@
+
+    private function configIsWritable() : bool
+    {
+        if (! file_exists($this->configFile) && ! is_writable(dirname($this->configFile))) {
```
I think we can simplify this function, according to the php docs:

> Returns TRUE if the filename exists and is writable. 
http://php.net/is_writable

So it's not enough just `return is_writable($this->configFile)` ?

(it checks existence)

weierophinney replied:
Noted; thanks!

weierophinney replied:
Actually, this only applies to the following conditional, not the one flagged. Note that this one is testing explicitly that the _parent directory_ is writable when the file does not exist.

The one following we can rewrite to `! is_writable($this->configFile)`, as we're then testing that it both exists and is writable.

weierophinney replied:
Okay, interestingly: tests FAIL when I change this! I'm not sure if this is a difference with how vfsStream handles the `is_writable()` call, or if the documented behavior of `is_writable` may be incorrect. Regardless, it needs to stay as-is.

michalbundyra replied:
Ok, I can check it later, for now it's fine for me. 👍

michalbundyra commented on src/Factory/FactoryAlreadyExistsException.php:
```diff
@@ -0,0 +1,24 @@
+
+use RuntimeException;
+
+class FactoryAlreadyExistsException extends RuntimeException
```
I would extract all Exceptions to namespace, as we have in different project.
I can see we don't have it here in commands, but we can refactor it later on as well. 
Here we have more exceptions so it will be clearer to have them in separate directory/namespace (each for command ofc)

weierophinney replied:
I'd rather not. Most of these commands are:

- A command class
- One or more utility classes to which they delegate
- One or more exceptions

The exceptions are specific to the specific command domain, and having a separate namespace is overkill, even when there are multiple different types. Consdiering the majority have a single exception type, forcing each to have a subnamespace is imposing hierarchy just for the sake of it.

michalbundyra commented on src/Factory/FactoryClassGenerator.php:
```diff
@@ -0,0 +1,135 @@
+    public function createFactory(string $className) : string
+    {
+        $class = $this->getClassName($className);
+        $namespace = str_replace('\\' . $class, '', $className);
```
Hm.. I can find a case when it will fail. I know, edge case, but I would use regexp or `strrpos` and `substr`

weierophinney replied:
Can you give me an example? Because I don't want to make a change unless there's a demonstrable error.

michalbundyra replied:
`App\EdgeName\Edge`

we have:
`$class = 'Edge'`
`$className = 'App\EdgeName\Edge'`,
and then:
`$namespace = 'AppName'`; what is obviously wrong.

michalbundyra commented on src/Factory/FactoryClassGenerator.php:
```diff
@@ -0,0 +1,135 @@
+    {
+        $reflectionClass = new ReflectionClass($className);
+
+        if (! $reflectionClass || ! $reflectionClass->getConstructor()) {
```
Hm... when we can have satisfied `! $reflectionClass`? I  can see that `ReflectionClass` can just throw exception on failure.

weierophinney replied:
You know, I have no idea. This was from the zend-servicemanager implementation, and, while I assume there was a reason for this, the PHP docs do indicate an exception is thrown normally. I'll update.

michalbundyra commented on src/Factory/FactoryClassGenerator.php:
```diff
@@ -0,0 +1,135 @@
+
+        $constructorParameters = $reflectionClass->getConstructor()->getParameters();
+
+        if (empty($constructorParameters)) {
```
we can use `! $constructorParameters` instead of `empty`.

michalbundyra commented on src/Factory/FactoryClassGenerator.php:
```diff
@@ -0,0 +1,135 @@
+            }
+        );
+
+        if (empty($constructorParameters)) {
```
as above

michalbundyra commented on src/Factory/FactoryClassGenerator.php:
```diff
@@ -0,0 +1,135 @@
+                }
+
+                if (null === $argument->getClass()) {
+                    throw new InvalidArgumentException(sprintf(
```
undefined class, missing import ?

michalbundyra commented on src/Factory/FactoryClassGenerator.php:
```diff
@@ -0,0 +1,135 @@
+            $constructorParameters,
+            function (ReflectionParameter $argument) {
+                if ($argument->isOptional()) {
+                    return false;
```
Do I understand correctly -  optional params are not going to be initialized in the factory?

weierophinney replied:
Correct. This is also how zend-servicemanager handles optional arguments in its `ReflectionBasedAbstractFactory` and related factory generator.

michalbundyra commented on test/CreateHandler/CreateHandlerCommandTest.php:
```diff
@@ -43,6 +54,32 @@ private function reflectExecuteMethod()
+
+        return $application;
+
+        $this->command->setApplication($application->reveal());
```
unreachable statement

weierophinney replied:
Oh! That's not good! I'll remove the `return` statement!

weierophinney replied:
Actually, need to remove the `setApplication()` statements, as they happen in the test cases that call this method.

michalbundyra commented on test/CreateHandler/CreateHandlerCommandTest.php:
```diff
@@ -43,6 +54,32 @@ private function reflectExecuteMethod()
         return $r;
     }
 
+    private function mockApplication()
```
As we are using PHP 7.1 I would suggest to add return type (`Application`?)

weierophinney replied:
Sure; do we need to add RTH within a test case, though?

michalbundyra replied:
We don't need to, but it would be nice, especially for IDEs and writing new test.
I would add docblock with `Application|ObjectProphecy` as it is more accurate, and maybe we should go this way, not RTH.

weierophinney replied:
This is... confusing. We can't specify a RTH of `Application`, as we actually return a prophecy. But specifying `ObjectProphecy` as the RTH makes it less clear what we're returning. I'll document it via a docblock annotation instead.

michalbundyra commented on test/CreateMiddleware/CreateMiddlewareCommandTest.php:
```diff
@@ -43,6 +54,32 @@ private function reflectExecuteMethod()
+
+        return $application;
+
+        $this->command->setApplication($application->reveal());
```
unreachable statement

weierophinney replied:
Will fix; thanks!

michalbundyra commented on test/Factory/FactoryClassGeneratorTest.php:
```diff
@@ -0,0 +1,53 @@
+        $className = InvokableObject::class;
+        $factory = file_get_contents(__DIR__ . '/TestAsset/factories/InvokableObject.php');
+
+        self::assertEquals($factory, $this->generator->createFactory($className));
```
Just for consistency, everywhere in this PR you are using `$this->` ... and maybe we can make a final decision what we are gonna do: `self::` or `$this->` (or `static::` 🤣 )

weierophinney replied:
I'll fix that; not sure why I used `self::` here; likely copy-pasta from the original in zend-servicemanager.

michalbundyra reviewed (commented):
It looks like that we are also missing definition of `const HELP_ARG_MODULE` in `CreateHandlerCommand` and `CreateFactoryCommand`.
For reference please see other command and class `final class CommandCommonOptions` (line 32)

**update:** oh, we are not calling `addDefaultOptionsAndArguments` so we don't need to define that constant there.

weierophinney committed:
Refactor `flag && require` statements to conditional blocks

weierophinney committed:
Remove parens around `include` argument

weierophinney committed:
Fix `mockApplication` method implementations

- Adds a docblock with the RTH specified as `ObjectProphecy|Application`
- Removes `$this->command->setApplication()` calls within the
  `mockApplication()` methods; these were unreachable, and unwanted.

weierophinney committed:
s/self::/$this->/ for consistency in test cases

weierophinney committed:
Better conditionals for FactoryClassGenerator

- Do not test if a `ReflectionClass` was produced; an exception is
  thrown if one is not.
- Use `! $constructorParameters` instead of `empty($constructorParameters)`
- An exception class was not imported, due to bad copy-pasta; created a
  new one with a named constructor, and used that instead.

michalbundyra commented on src/Factory/FactoryClassGenerator.php:
```diff
@@ -0,0 +1,131 @@
+        return substr($className, strrpos($className, '\\') + 1);
+    }
+
+    private function getConstructorParameters(string $className) : array
```
Missing `@throws` annotation with the function (`UnidentifiedTypeException`)

michalbundyra commented on src/Factory/ConfigInjector.php:
```diff
@@ -0,0 +1,115 @@
+    /**
+     * Sorts entries by key, using natcase
+     */
+    private function sort(array $config) : array
```
This is basically `ksort` with `SORT_NATURAL` flag :) So I would suggest to use internal function instead.

weierophinney committed:
Fixes how FactoryClassGenerator determines namespace from submitted class

Uses `substr` + `strrpos` to calculate it.

weierophinney committed:
Simplifies `ConfigInjection::configIsWritable()` logic

weierophinney committed:
Adds a missing `@throws` annotation

weierophinney committed:
Use `ksort()` with the `SORT_NATURAL` sort flag, instead of homegrown implementation.
